Design/3093-graphql-defer.md

```diff
@@ -0,0 +1,356 @@
+    completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
+  ) where Operation : GraphQLOperation {
+    // response validators not shown
+
+    guard
+      let multipartBoundary = response.httpResponse.multipartBoundary,
+      let protocolSpec = response.httpResponse.multipartProtocolSpec,
+      let protocolParser = Self.multipartParsers[protocolSpec]
+    else {
+      // call request chain error handler
+
+      return
+    }
+
+    let dataHandler: ((Data) -> Void) = { data in
+      // proceed ahead on the request chain
+    }
+
+    let errorHandler: (() -> Void) = {
+      // call request chain error handler
+    }
+
+    protocolParser.parse(
+      data: response.rawData,
+      boundary: multipartBoundary,
+      dataHandler: dataHandler,
+      errorHandler: errorHandler
+    )
+  }
+}
+
+// Sample protocol for multipart specification parsing
+protocol MultipartResponseSpecificationParser {
+  static var protocolSpec: String { get }
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  )
+}
+
+// Sample implementations of multipart specification parsers
+
+struct SubscriptionResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "subscriptionSpec=1.0"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // parsing code currently in MultipartResponseParsingInterceptor
+  }
+}
+
+struct DeferResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "deferSpec=20220824"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // new code to parse the defer specification
+  }
+}
+```
+
+#### Response data
+
+The initial response data and data received in each incremental response will need to be retained and combined so that each incremental response can insert the latest received incremental response data at the correct path and return an up-to-date response to the request callback.
+
+The data being retained and combined should not require another pass through the GraphQL executor though.
+
+### Completion handler
+
+`GraphQLResult` should be modified to provide query completion blocks with a high-level abstraction of whether the request has been fulfilled or is still in progress. This prevents clients from having to dig into the deferred fragments to identify the state of the overall request.
+
+Potential solutions:
+1. Introduce a new property on the `GraphQLResult` type that can be used to express the state of the request.
+
+```swift
+// New Response type and property
+public struct GraphQLResult<Data: RootSelectionSet> {
+  // other properties and types not shown
+
+  public enum Response {
+    case partial
+    case complete
+  }
+
+  public let response: Response
```

Comment:
Oh, this is what I was referring to above! :) 

I don't like calling this `Response` though. It's not actually the response data, it's just indicating if the response is partial/complete. Think we just need some consideration on naming here.

Reply:
Noted

Reply:
I do like the idea of providing this data to the user through an enum. As for naming, yea just `Response` maybe isn't the best, thinking about what we are trying to represent which is the status of the result, maybe something like `ResultState`? `ResponseResult` doesn't really feel right, and `ResultData` would make me think it is data and not a state representation.

Reply:
The naming we can settle on in Slack or the PR. It seems like we're aligned on the state being available at the root of `GraphQLResult` and not only on the `server` source case.